src/charm.py

```diff
@@ -979,6 +1057,118 @@ def auth_enabled(self) -> bool:
 
         return any("MONGOD_ARGS" in line and "--auth" in line for line in env_vars)
 
+    def _scope_obj(self, scope: Scopes):
+        if scope == APP_SCOPE:
+            return self.app
+        if scope == UNIT_SCOPE:
+            return self.unit
+
+    def _peer_data(self, scope: Scopes):
+        if not self._peers:
+            return {}.setdefault(scope, {})
+        scope_obj = self._scope_obj(scope)
+        return self._peers.data[scope_obj]
+
+    def _juju_secret_set(self, scope: Scopes, key: str, value: str) -> str:
+        """Helper function setting Juju secret."""
+        peer_data = self._peer_data(scope)
+        self._juju_secrets_get(scope)
+
+        secret = self.secrets[scope].get(Config.Secrets.SECRET_LABEL)
+
+        # It's not the first secret for the scope, we can re-use the existing one
+        # that was fetched in the previous call
+        if secret:
+            secret_cache = self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL]
+
+            if secret_cache.get(key) == value:
+                logging.debug(f"Key {scope}:{key} has this value defined already")
+            else:
+                secret_cache[key] = value
+                try:
+                    secret.set_content(secret_cache)
+                    logging.debug(f"Secret {scope}:{key} was {key} set")
+                except OSError as error:
+                    logging.error(
+                        f"Error in attempt to set '{key}' secret for scope '{scope}'. "
+                        f"Existing keys were: {list(secret_cache.keys())}. {error}"
+                    )
+
+        # We need to create a brand-new secret for this scope
+        else:
+            scope_obj = self._scope_obj(scope)
+
+            secret = scope_obj.add_secret({key: value})
+            if not secret:
+                raise SecretNotAddedError(f"Couldn't set secret {scope}:{key}")
+
+            self.secrets[scope][Config.Secrets.SECRET_LABEL] = secret
+            self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL] = {key: value}
+            logging.debug(f"Secret {scope}:{key} published (as first). ID: {secret.id}")
+            peer_data.update({Config.Secrets.SECRET_INTERNAL_LABEL: secret.id})
+
+        return self.secrets[scope][Config.Secrets.SECRET_LABEL].id
+
+    def _juju_secrets_get(self, scope: Scopes) -> Optional[bool]:
```

Comment:
seems like some functions rely on the result of `_juju_secrets_get `, should it really return an optional bool?

Reply:
no, it should not, and function naming is a bit misleading in my opinion, but for this PR let's keep it as-is because it was copied from k8s charm in that way and will require changes in both charms